src/webgpu/api/operation/rendering/draw.spec.ts

```diff
@@ -44,7 +28,330 @@ TODO:
+The test checks that the center of all of the expected triangles is drawn, and the others are empty.
+The fragment shader also writes out to a storage buffer. If the draw is zero-sized, check that no value is written.
+
+Params:
+  - first= {0, 3} - either the firstVertex or firstIndex
+  - count= {0, 6} - either the vertexCount or indexCount
+  - first_instance= {0, 2}
+  - instance_count= {0, 4}
+  - indexed= {true, false}
+  - indirect= {true, false}
+  - vertex_buffer_offset= {0, 32}
+  - index_buffer_offset= {0, 16} - only for indexed draws
+  - base_vertex= {0, 9} - only for indexed draws
+  `
+  )
+  .cases(
+    params()
+      .combine(poptions('first', [0, 3] as const))
+      .combine(poptions('count', [0, 6] as const))
+      .combine(poptions('first_instance', [0, 2] as const))
+      .combine(poptions('instance_count', [0, 4] as const))
+      .combine(pbool('indexed'))
+      .combine(pbool('indirect'))
+      .combine(poptions('vertex_buffer_offset', [0, 32] as const))
+      .expand(p => poptions('index_buffer_offset', p.indexed ? ([0, 16] as const) : [undefined]))
+      .expand(p => poptions('base_vertex', p.indexed ? ([0, 9] as const) : [undefined]))
+  )
+  .fn(t => {
+    const renderTargetSize = [36, 36];
+
+    // The test will split up the render target into a grid where triangles of
+    // increasing primitive id will be placed along the X axis, and triangles
+    // of increasing instance id will be placed along the Y axis. The size of the
+    // grid is based on the max primitive id and instance id used.
+    const numX = 3;
+    const numY = 6;
+    const tileSizeX = renderTargetSize[0] / numX;
+    const tileSizeY = renderTargetSize[1] / numY;
+
+    // |\
+    // |   \
+    // |______\
+    // Unit triangle shaped like this. 0-1 Y-down.
+    const bottomLeftTriangleVertices = /* prettier-ignore */ [
+      0.0, 0.0,
+      0.0, 1.0,
+      1.0, 1.0,
+    ];
+
+    // _______
+    // \      |
+    //    \   |
+    //       \|
+    // Unit triangle shaped like this. 0-1 Y-down.
+    const topRightTriangleVertices = /* prettier-ignore */ [
+      0.0, 0.0,
+      1.0, 1.0,
+      1.0, 0.0,
+    ];
+
+    const renderTarget = t.device.createTexture({
+      size: renderTargetSize,
+      usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC,
+      format: 'rgba8unorm',
+    });
+
+    const vertexModule = t.device.createShaderModule({
+      code: `
+[[builtin(vertex_index)]] var<in> vertex_index : u32;
+[[builtin(instance_index)]] var<in> instance_id : u32;
+[[location(0)]] var<in> vertexPosition : vec2<f32>;
+
+[[builtin(position)]] var<out> Position : vec4<f32>;
+[[stage(vertex)]] fn vert_main() -> void {
+  // 9u is the base_vertex offset. We select different vertices
+  // but apply the same x offset in the grid.
+  // 3u is the number of points in a triangle to convert from index
+  // to id.
+  var vertex_id : u32 = (vertex_index % 9u) / 3u;
```

Comment:
It could be possible to use a larger render target to now have to mess with the best vertex in the shader.

Reply:
Done.